### PR TITLE
Replace "import *" with explicit imports in numba/types

### PR DESCRIPTION
**Pull request by asmeurer**
There aren't any other files that "import *" from these files, so there
shouldn't be any breakage due to side effects.

c.f. #2952

Also when fixing this, I found two undefined variables in these files.

* `intp` is undefined here: https://github.com/numba/numba/blob/37ab1420fa87d02b635991b32f0faf5c65e8fc4d/numba/types/containers.py#L179
* `this` is undefined here: https://github.com/numba/numba/blob/37ab1420fa87d02b635991b32f0faf5c65e8fc4d/numba/types/functions.py#L90

I'm unsure how to fix these so I've left them be for now. 

Undefined variables like these are easy to find with pyflakes, but only if you don't have star imports!

I may also look at removing star imports in other files. I wrote a little tool that makes replacing star imports with explicit imports easy https://github.com/asmeurer/removestar.

### CODE REVIEW

**Comment by esc** [comment]
@asmeurer thanks for cleaning this up, it is much appreciated. The `intp` might be the following:

https://github.com/numba/numba/blob/37ab1420fa87d02b635991b32f0faf5c65e8fc4d/numba/types/__init__.py#L48


The `this` might be a legit bug.

**Comment by asmeurer** [comment]
The getitem method is clearly never used, since it raises an exception (I double checked master and it does indeed give a NameError with `intp`). I'm unclear what it's supposed to be. `__getitem__` is also defined right below it, returning `self.dtype`. I'm loathe to fix it without understanding what it really should be. It seems it could also be a typo for `ind`, the name of the second argument of the method. It was added in this commit 2518106b53, where it does indeed seem to be what you are suggesting. But it's also used in some other places where it is not used in master. So maybe `getitem` was removed at some point in favor of `__getitem__`, but this one was missed? 

Regarding `this`, it was added in this commit: 783a2de4eb. It isn't clear to me from that commit what it is supposed to be.

**Comment by esc** [comment]
It seems tricky to figure out how to deal with these two 'bugs'. The have been in there for quite some time and probably haven't caused any issues, I would assume. So probably it is safe to just delete them? Maybe we should do exactly that and then send the commit through the build farm to see if it breaks anything?

**Comment by stuartarchibald** [comment]
Think the `getitem()` call is a historical artefact from initial getitem impls https://github.com/numba/numba/commit/2518106b53c46e7d5605f266be25071144ce2e1a#diff-0631947c9fcac56e361a815559a25b50R349-R365, I think it can safely be removed.

I think `this` should probably be `keys` from here https://github.com/numba/numba/blob/37ab1420fa87d02b635991b32f0faf5c65e8fc4d/numba/types/functions.py#L87.

**Comment by asmeurer** [comment]
Both seems correct to me as well looking at the code/history.

**Comment by asmeurer** [comment]
I've pushed a commit fixing them. I don't know if a test should be added for the `this` exception.

**Comment by asmeurer** [comment]
I also went ahead and fixed the remaining pyflakes errors in that submodule (removed unused imports).

File: numba/types/functions.py [hunk 1]
```diff
@@ -87,7 +84,7 @@ def __init__(self, template):
             keys = set(temp.key for temp in self.templates)
             if len(keys) != 1:
                 raise ValueError("incompatible templates: keys = %s"
-                                 % (this,))
+                                 % (keys,))
```
Comment (stuartarchibald) [inline numba/types/functions.py]: This snippet may be useful to check that this is now working as expected:
```python
from numba.typing.templates import make_overload_template
from numba import types

def func_stub():
    pass

def func_stub2():
    pass

def ol():
    pass

template1 = make_overload_template(func_stub, ol, {}, True)
template2 = make_overload_template(func_stub2, ol, {}, True)

types.Function((template1, template2)) # should fail
```

Comment (jrbourbeau) [reply numba/types/functions.py]: I can confirm the above snippet gives

```
NameError: name 'this' is not defined
```

on `master`, and 

```
ValueError: incompatible templates: keys = {<function func_stub2 at 0x10d236158>, <function func_stub at 0x10ce720d0>}
```

with this PR

**Comment by asmeurer** [comment]
OK, I added a test, and went ahead and removed the star import in the test file as well. This is ready to go, unless you want me to remove star imports anywhere else.

File: numba/tests/test_types.py [hunk 2]
```diff
@@ -217,6 +218,22 @@ def test_list_type_getitem(self):
...
+            pass
+
+        def ol():
+            pass
+
+        template1 = make_overload_template(func_stub, ol, {}, True)
+        template2 = make_overload_template(func_stub2, ol, {}, True)
+
+        with self.assertRaises(ValueError):
+            types.Function((template1, template2))
```
Comment (stuartarchibald) [inline numba/tests/test_types.py]: Think this needs another new line here, should be two between class definitions?

File: numba/tests/test_types.py [hunk 3]
```diff
@@ -217,6 +218,22 @@ def test_list_type_getitem(self):
...
+        def func_stub2():
+            pass
+
+        def ol():
+            pass
+
+        template1 = make_overload_template(func_stub, ol, {}, True)
+        template2 = make_overload_template(func_stub2, ol, {}, True)
+
+        with self.assertRaises(ValueError):
```
Comment (stuartarchibald) [inline numba/tests/test_types.py]: There's a preference in new code to use the following pattern to check that the exception is the exception expected (`ValueError` and `TypingError` appear 90% of the time and it's very easy to get them from an unexpected place).

```python
with self.assertRaises(ValueError) as raises:
    types.Function((template1, template2))
self.assertIn("incompatible templates:", str(raises.exception))
```

File: numba/tests/test_types.py [hunk 4]
```diff
@@ -18,10 +18,11 @@
 from numba.utils import IS_PY3
 from numba import sigutils, types, typing
 from numba.types.abstract import _typecache
+from numba.typing.templates import make_overload_template
 from numba import jit, numpy_support, typeof
 from numba.numpy_support import version as numpy_version
```
Comment (stuartarchibald) [inline numba/tests/test_types.py]: Whilst we're fixing up imports, I think this import is unused?

File: numba/types/containers.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from __future__ import print_function, division, absolute_import
 
-from .abstract import *
-from .common import *
+from .abstract import ConstSized, Container, Hashable, MutableSequence, Sequence, Type, TypeRef
```
Comment (stuartarchibald) [inline numba/types/containers.py]: Line >80 chars.

File: numba/types/functions.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/types/functions.py]: Thanks for confirming @jrbourbeau

**Comment by asmeurer** [comment]
I've made the requested changes. I can remove `import *` from other places in numba as well if you like. Otherwise, this is ready to go.

**Comment by asmeurer** [comment]
Here's the other import *'s that could be removed. I haven't tested if removing them breaks any tests. Also note the warnings https://gist.github.com/asmeurer/404afe6d8e115ae8837a83616502c925

**Comment by stuartarchibald** [comment]
> Here's the other import *'s that could be removed. I haven't tested if removing them breaks any tests. Also note the warnings https://gist.github.com/asmeurer/404afe6d8e115ae8837a83616502c925

Additional fixes along these lines would be welcome. I think some of those warnings are from detecting that the exact same things are imported via `*` from different modules, fixing this too would be great. Thanks.

**Comment by asmeurer** [comment]
It looks like a lot of those `import *`s are legitimate importing of namespaces. I need to weed out the ones that break tests. removestar currently doesn't know how to check if import *'d names are used somewhere else. 

There's also an error because test_multi3.py has a Unicode BOM character at the beginning of the file, which is technically a bug in my code, but I suppose that character probably shouldn't be there.

**Comment by stuartarchibald** [comment]
> It looks like a lot of those `import *`s are legitimate importing of namespaces. I need to weed out the ones that break tests. removestar currently doesn't know how to check if import *'d names are used somewhere else.
> 
> There's also an error because test_multi3.py has a Unicode BOM character at the beginning of the file, which is technically a bug in my code, but I suppose that character probably shouldn't be there.

Great, thanks. What you describe that needs fixing seems about right, if you choose to implement these changes please feel free to open a new PR, this current one will be merged during the next merge window. Thanks again!

**Comment by seibert** [comment]
It seems that one of the PRs I merged before this one has created a merge conflict.  Can you merge in master and resolve the conflict?

**Comment by asmeurer** [comment]
Done

**Comment by stuartarchibald** [comment]
There's currently a test fail on CI when merged with updated master:
```
======================================================================
ERROR: test_function_incompatible_templates (numba.tests.test_types.TestTypes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "numba/tests/test_types.py", line 231, in test_function_incompatible_templates
    template1 = make_overload_template(func_stub, ol, {}, True)
TypeError: make_overload_template() takes exactly 5 arguments (4 given)

----------------------------------------------------------------------
```
the function in question is now like:
```python
def make_overload_template(func, overload_func, jit_options, strict, inline):
```
since `inline` was added to the `@overload` decorator.


@asmeurer I've made a PR against your fork for this branch https://github.com/asmeurer/numba/pull/1 that should fix it.

**Comment by stuartarchibald** [comment]
@asmeurer that was quick, thanks for the merge. Hopefully that'll make CI happy!

**Pull request merged**